File: generate-world-from-text.py
"""Generate a World Labs world from a text prompt."""

import json
import os
import sys
import time
from urllib import request, error


# Get World Labs API key from environment variable
WLT_API_KEY = os.environ.get("WLT_API_KEY")
if not WLT_API_KEY:
    print("Please set environment variable WLT_API_KEY to your World Labs API key")
    sys.exit(1)

API_BASE_URL = "http://api-autopush.worldlabs.ai/marble/v1"


def api_fetch(path, method="GET", body=None, headers=None):
    """Send an API request and parse JSON response."""
    url = f"{API_BASE_URL}/{path}"
    payload = json.dumps(body).encode("utf-8") if body is not None else None
    req_headers = {"WLT-Api-Key": WLT_API_KEY, "Content-Type": "application/json"}
    if headers:
        req_headers.update(headers)
    req = request.Request(url, data=payload, headers=req_headers, method=method)
    try:
        with request.urlopen(req) as response:
            response_body = response.read().decode("utf-8")
            return json.loads(response_body) if response_body else {}
    except error.HTTPError as exc:
        response_body = exc.read().decode("utf-8")
        raise RuntimeError(
            f"{path}: {exc.code} {response_body}"
        ) from exc


def generate_world(text_prompt, auto_enhance, draft, seed):
    """Start a world generation operation from text."""
    if not text_prompt:
        raise RuntimeError("Enter a text prompt first.")

    world_prompt = {
        "type": "text",
        "text_prompt": text_prompt,
        "disable_recaption": not auto_enhance,
    }
    operation = api_fetch(
        "worlds:generate",
        method="POST",
        body={
            "world_prompt": world_prompt,
            "model": "Marble 0.1-mini" if draft else "Marble 0.1-plus",
            "seed": seed,
        },
    )
    return operation["operation_id"]


def get_operation(operation_id):
    """Fetch operation status by ID."""
    return api_fetch(f"operations/{operation_id}")


def get_world(world_id):
    """Fetch world details by ID."""
    return api_fetch(f"worlds/{world_id}")


def main():
    """Run the CLI flow for text-based generation."""
    if len(sys.argv) < 2:
        print("Usage: python generate-world-from-text.py 'A cozy modern interior jazz lounge'")
        sys.exit(1)

    text_prompt = sys.argv[1]
    auto_enhance = True
    draft = True
    seed = None

    operation_id = generate_world(text_prompt, auto_enhance, draft, seed)
    print(f'Submitted world generation for "{text_prompt}"', {"operationId": operation_id})

    operation = None
    while True:
        operation = get_operation(operation_id)
        if operation.get("done"):
            break
        print(f"Operation {operation_id} still processing")
        time.sleep(5)

    world_id = operation["response"]["world_id"]
    world = get_world(world_id)
    print("World", json.dumps(world, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: generate-world-from-text.js
// Generate a World Labs world from a text prompt.

// Get World Labs API key from environment variable
const WLT_API_KEY = process.env.WLT_API_KEY;
if (!WLT_API_KEY) {
  console.error("Please set environment variable WLT_API_KEY to your World Labs API key");
  process.exit(1);
}

const apiBaseUrl = "https://api.worldlabs.ai/marble/v1";

// Send an API request and parse JSON response.
async function apiFetch(path, options = {}) {
  const response = await fetch(`${apiBaseUrl}/${path}`, {
    ...options,
    headers: {
      "WLT-Api-Key": WLT_API_KEY,
      "Content-Type": "application/json",
      ...options.headers,
    },
  });
  const body = await response.json();
  if (!response.ok) {
    throw new Error(`${path}: ${response.status} ${JSON.stringify(body)}`);
  }
  return body;
}

// Start a world generation operation from text.
async function generateWorld(textPrompt, autoEnhance, draft, seed) {
  if (!textPrompt) {
    throw new Error("Enter a text prompt first.");
  }

  const worldPrompt = {
    type: "text",
    text_prompt: textPrompt,
    disable_recaption: !autoEnhance,
  };
  const operation = await apiFetch("worlds:generate", {
    method: "POST",
    body: JSON.stringify({
      world_prompt: worldPrompt,
      model: draft ? "Marble 0.1-mini" : "Marble 0.1-plus",
      seed,
    }),
  });
  return operation.operation_id;
}

// Fetch operation status by ID.
async function getOperation(operationId) {
  return await apiFetch(`operations/${operationId}`);
}

// Fetch world details by ID.
async function getWorld(worldId) {
  return await apiFetch(`worlds/${worldId}`);
}


// Run the CLI flow for text-based generation.
const textPrompt = process.argv[2];
if (!textPrompt) {
  console.error("Usage: node generate-world-from-text.js 'A cozy modern interior jazz lounge'");
  process.exit(1);  
}

const autoEnhance = true;
const draft = true;
const seed = null;

const operationId = await generateWorld(textPrompt, autoEnhance, draft, seed);
console.log(`Submitted world generation for "${textPrompt}"`, { operationId });

let operation;
while (true) {
  operation = await getOperation(operationId);
  if (operation.done) {
    break;
  }
  
  console.log(`Operation ${operationId} still processing`);
  // Wait 5000ms before trying again
  await new Promise((resolve) => setTimeout(resolve, 5000));
}

const worldId = operation.response.world_id;
const world = await getWorld(worldId);
console.log("World", world);
